plat/allwinner/common/sunxi_cpu_ops.c

```diff
@@ -45,25 +45,27 @@ static void sunxi_cpu_enable_power(unsigned int cluster, unsigned int core)
 	mmio_write_32(SUNXI_CPU_POWER_CLAMP_REG(cluster, core), 0x00);
 }
 
-void sunxi_cpu_off(unsigned int cluster, unsigned int core)
+void sunxi_cpu_off(u_register_t mpidr)
 {
-	int corenr = cluster * PLATFORM_MAX_CPUS_PER_CLUSTER + core;
+	unsigned int cluster = MPIDR_AFFLVL1_VAL(mpidr);
+	unsigned int core    = MPIDR_AFFLVL0_VAL(mpidr);;
 
 	VERBOSE("PSCI: Powering off cluster %d core %d\n", cluster, core);
 
 	/* Deassert DBGPWRDUP */
 	mmio_clrbits_32(SUNXI_CPUCFG_DBG_REG0, BIT(core));
 
 	/* We can't turn ourself off like this, but it works for other cores. */
-	if (plat_my_core_pos() != corenr) {
+	if (read_mpidr() != mpidr) {
 		/* Activate the core output clamps, but not for core 0. */
-		if (corenr != 0)
+		if (core != 0)
 			mmio_setbits_32(SUNXI_POWEROFF_GATING_REG(cluster),
-					BIT(core));
+			                BIT(core));
 		/* Assert CPU power-on reset */
 		mmio_clrbits_32(SUNXI_POWERON_RST_REG(cluster), BIT(core));
-		/* Remove power from the CPU */
-		sunxi_cpu_disable_power(cluster, core);
+		/* Remove power from the CPU, but not for core 0. */
+		if (core != 0)
+			sunxi_cpu_disable_power(cluster, core);
```

Comment:
Interesting, how did you trigger this issue? A PSCI SYSTEM_OFF from a core other than core 0?
And shouldn't this hit in the OpenRISC code as well? I am pretty sure I offlined core 0 multiple times during my randomized offlining test.

Reply:
> And is there a way to do a patch-by-patch review in Github? This whole PR-at-once process is quite painful ...

Yes, you can go to the "Commits" tab, click on one, and start reviewing inline from there. Though as far as I know, GitHub still likes to get the commits in the wrong order in PRs.

Reply:
Funny you should mention that :). No, it can't be (easily) triggered with the current code. `sunxi_cpu_off` is only called from 3 places:
* `sunxi_disable_secondary_cpus` from `bl31_early_platform_setup2`, which is always called from CPU0, so ignores CPU0
* `sunxi_pwr_down_wfi`, which is always called from the CPU being shutdown, so it hangs at the previous line.
* `sunxi_disable_secondary_cpus` from `sunxi_system_off`, which will only be called from CPU0 (Linux's "boot" CPU) if that CPU is online. And if that CPU is already offline... actually, let me check if it is hit in that case.

So you would only hit it when performing a poweroff, and you wouldn't notice that the CPU hung, because that's what it's supposed to do. But if you look at the A64 manual, the diagram in CPUCFG (section 3.4.2, page 148), there's no power switch for core 0. That's `DCDC_EN` for the whole cluster. That's corroborated by the `R_PRCM` documentation for the A83T, which has the core 0 power switch in bit 4, and the cluster power switch in bit 0. And I've experimentally tested this from the ARISC. Search for `R_PRCM_PWROFF_GATING_REG` in [this file](https://github.com/crust-firmware/crust/blob/9f3157394ec430d2198441fe062d00e734a554ef/platform/sun50i/css.c).

Reply:
Also see references to `set_power_switch` in that file. And finally that's what Allwinner does in their blob wrt the power switch (which I can't link directly because there's a few layers of indirection involved in that function).

Reply:
Huh, well the call to `sunxi_disable_secondary_cpus` from `sunxi_system_off` doesn't crash, even with CPU0 offline, so I must have something mixed up. I reverted that change as well for now, since it's not necessary.